File: examples/positions.ts
/**
 * Floe Credit API — Portfolio inspection (FLO-530).
 *
 * Walks an agent through reading its credit portfolio via
 * GET /v1/positions/:wallet:
 *
 *   1. Indexer-mode read (default) — fast; fine for "what do I owe".
 *                                     `worstHealthFactor` is null because the
 *                                     indexer returns origination LTV which
 *                                     doesn't reflect price movement.
 *   2. Live-mode read    (?live=true) — slower, fully on-chain. Health factor
 *                                       is computed against the live oracle.
 *   3. includePending     — surface borrow intents that haven't matched yet.
 *   4. Pagination         — loop with ?skip= until activeLoanCount < limit.
 *   5. 503 handling       — what to do when the API instance has no indexer.
 *
 * Usage:
 *   npm install viem
 *   PRIVATE_KEY=0x... npx tsx positions.ts
 */

import { privateKeyToAccount } from "viem/accounts";
import type { Hex } from "viem";

// ── Config ──

const PRIVATE_KEY = process.env.PRIVATE_KEY as Hex;
if (!PRIVATE_KEY) {
  console.error("Set PRIVATE_KEY=0x... environment variable");
  process.exit(1);
}

const API_BASE = process.env.FLOE_API_BASE ?? "https://credit-api.floelabs.xyz";

// Wallet whose portfolio we want to read. Defaults to the signer's own.
// Any authenticated caller can query any wallet's positions today.
const TARGET_WALLET = process.env.TARGET_WALLET as `0x${string}` | undefined;

// Pagination — small for the example so the loop is exercised on
// realistic portfolios. Production callers can use the API's defaults.
const PAGE_LIMIT = 5;

// ── Setup ──

const account = privateKeyToAccount(PRIVATE_KEY);
const targetWallet = TARGET_WALLET ?? account.address;

interface PositionsSummary {
  activeLoanCount: number;
  totalPrincipal: string;
  totalDebt: string;
  totalCollateralLocked: string;
  weightedAvgInterestRateBps: string | null;
  worstHealthFactor: number | null;
  nextMaturityAt: string | null;
  perCollateralToken: Record<string, string>;
}

interface BorrowIntentSummary {
  offerHash: string;
  marketId: string;
  borrowAmount: string;
  collateralAmount: string;
  maxInterestRateBps: string;
  minLtvBps: string;
  expiry: string;
  status: string;
}

interface CreditPositions {
  account: string;
  positions: Array<{
    loanId: string;
    totalDebt: string;
    bufferBps: string;
    [k: string]: unknown;
  }>;
  summary: PositionsSummary;
  pendingBorrowIntents?: BorrowIntentSummary[];
  source: "indexer" | "chain";
  ltvStale: boolean;
  indexerBlockNumber: string | null;
  fetchedAt: number;
}

// ── Auth helper (mirrors examples/borrow.ts) ──

async function signRequest(): Promise<Record<string, string>> {
  // Mirrors the helper in examples/borrow.ts. Keep them in lockstep
  // when the message format changes.
  const timestamp = Math.floor(Date.now() / 1000).toString();
  const message = `Floe Credit API\nTimestamp: ${timestamp}`;
  const signature = await account.signMessage({ message });
  return {
    "X-Wallet-Address": account.address,
    "X-Signature": signature,
    "X-Timestamp": timestamp,
  };
}

interface GetPositionsOpts {
  live?: boolean;
  includePending?: boolean;
  limit?: number;
  skip?: number;
  pendingLimit?: number;
}

async function getPositions(
  wallet: string,
  opts: GetPositionsOpts = {},
): Promise<CreditPositions> {
  const url = new URL(`${API_BASE}/v1/positions/${wallet}`);
  if (opts.live) url.searchParams.set("live", "true");
  if (opts.includePending) url.searchParams.set("includePending", "true");
  if (opts.limit !== undefined) url.searchParams.set("limit", String(opts.limit));
  if (opts.skip !== undefined) url.searchParams.set("skip", String(opts.skip));
  if (opts.pendingLimit !== undefined) {
    url.searchParams.set("pendingLimit", String(opts.pendingLimit));
  }

  const resp = await fetch(url, { headers: await signRequest() });

  if (resp.status === 503) {
    // The API instance was started without an Envio indexer endpoint.
    // getPositions cannot run without it (chain-only fallback would
    // be too slow). Retry against a different instance or surface to
    // the operations team.
    throw new Error(
      `503 — indexer not configured on ${API_BASE}. Try a different API instance.`,
    );
  }
  if (!resp.ok) {
    throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  }
  return (await resp.json()) as CreditPositions;
}

function fmtRate(bps: string | null | undefined): string {
  if (bps == null) return "—";
  return `${(Number(bps) / 100).toFixed(2)}%`;
}

function fmtUsdc(raw: string | bigint): string {
  const n = typeof raw === "string" ? BigInt(raw) : raw;
  return `${(Number(n) / 1e6).toLocaleString(undefined, { maximumFractionDigits: 2 })} USDC`;
}

function fmtHealth(hf: number | null): string {
  if (hf === null) return "n/a (indexer mode)";
  let tag: string;
  if (hf >= 2.0) tag = "safe";
  else if (hf >= 1.2) tag = "ok";
  else if (hf > 1.0) tag = "tight";
  else tag = "LIQUIDATABLE";
  return `${hf.toFixed(3)} (${tag})`;
}

function printSummary(label: string, body: CreditPositions): void {
  const s = body.summary;
  console.log(`--- ${label} ---`);
  console.log(`  source              : ${body.source}`);
  console.log(`  ltvStale            : ${body.ltvStale}`);
  console.log(`  indexerBlockNumber  : ${body.indexerBlockNumber}`);
  console.log(`  active loans        : ${s.activeLoanCount}`);
  console.log(`  total debt          : ${fmtUsdc(s.totalDebt)}`);
  console.log(`  weighted-avg rate   : ${fmtRate(s.weightedAvgInterestRateBps)}`);
  console.log(`  worst health factor : ${fmtHealth(s.worstHealthFactor)}`);
  if (s.nextMaturityAt) {
    console.log(`  next maturity       : ${s.nextMaturityAt} (unix)`);
  }
  console.log();
}

async function main(): Promise<void> {
  console.log(`Signer:  ${account.address}`);
  console.log(`Target:  ${targetWallet}`);
  console.log(`API:     ${API_BASE}`);
  console.log();

  // ── Step 1: indexer-mode read ──────────────────────────────────────
  let body = await getPositions(targetWallet);
  printSummary("Step 1: indexer mode (default)", body);

  // The summary's worstHealthFactor is intentionally null in indexer
  // mode. Calling code that drives liquidation alarms should always
  // use live mode (Step 2).

  // ── Step 2: live-mode read ─────────────────────────────────────────
  body = await getPositions(targetWallet, { live: true });
  printSummary("Step 2: live mode (?live=true)", body);

  // ── Step 3: include pending borrow intents ─────────────────────────
  body = await getPositions(targetWallet, { includePending: true });
  const pending = body.pendingBorrowIntents ?? [];
  console.log(`--- Step 3: pending borrow intents (${pending.length}) ---`);
  for (const intent of pending.slice(0, 5)) {
    console.log(
      `  • ${intent.offerHash}  ` +
        `borrow=${fmtUsdc(intent.borrowAmount)}  ` +
        `max=${fmtRate(intent.maxInterestRateBps)}  ` +
        `expires=${intent.expiry}`,
    );
  }
  if (pending.length > 5) {
    console.log(`  … and ${pending.length - 5} more`);
  }
  console.log();

  // ── Step 4: paginate ───────────────────────────────────────────────
  // Per the API contract, paginate as long as `loans.length === limit`
  // (the page is "full"). Don't gate on summary.activeLoanCount —
  // that's the wallet-wide total and stays constant across pages.
  console.log(`--- Step 4: paginate active loans (page size ${PAGE_LIMIT}) ---`);
  let seen = 0;
  let skip = 0;
  while (true) {
    const page = await getPositions(targetWallet, { limit: PAGE_LIMIT, skip });
    const loans = page.positions;
    for (const loan of loans) {
      seen += 1;
      console.log(
        `  loan #${loan.loanId}: ` +
          `debt=${fmtUsdc(loan.totalDebt)}, ` +
          `buffer=${fmtRate(loan.bufferBps)}`,
      );
    }
    if (loans.length < PAGE_LIMIT) break;
    skip += PAGE_LIMIT;
  }
  console.log(`  total surfaced via pagination: ${seen}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/positions.py
"""
Floe Credit API — Portfolio inspection (FLO-530, SELF-CUSTODY).

⚠️ This example uses a raw PRIVATE_KEY for wallet-signature auth. The
   managed-wallet flow (recommended for nearly all agents) authenticates
   with a `floe_…` runtime key instead — no key custody on your side. See
   docs/getting-started/quickstart.md and docs/developers/self-custody.md.


Walks an agent through reading its credit portfolio via
GET /v1/positions/:wallet:

  1. Indexer-mode read (default) — fast; fine for "what do I owe".
                                    `worstHealthFactor` is null because the
                                    indexer returns origination LTV which
                                    doesn't reflect price movement.
  2. Live-mode read    (?live=true) — slower, fully on-chain. Health factor
                                      is computed against the live oracle.
  3. includePending     — surface borrow intents that haven't matched yet.
  4. Pagination         — loop with ?skip= until positions.length < limit.
  5. 503 handling       — what to do when the API instance has no indexer.

Usage:
  pip install web3 requests
  PRIVATE_KEY=0x... python positions.py
"""

import os
import re
import sys
import time
from urllib.parse import quote

import requests
from eth_account import Account
from eth_account.messages import encode_defunct

# Defense-in-depth: reject obviously malformed wallets before we
# interpolate them into the URL path. The API also validates, but a
# caller-side check produces a clearer error and avoids a wasted round
# trip if TARGET_WALLET is misconfigured.
ADDRESS_RE = re.compile(r"^0x[a-fA-F0-9]{40}$")

# ── Config ──

PRIVATE_KEY = os.environ.get("PRIVATE_KEY")
if not PRIVATE_KEY:
    print("Set PRIVATE_KEY=0x... environment variable")
    sys.exit(1)

API_BASE = os.environ.get("FLOE_API_BASE", "https://credit-api.floelabs.xyz")

# Wallet whose portfolio we want to read. Defaults to the signer's own.
# Any authenticated caller can query any wallet's positions today.
TARGET_WALLET = os.environ.get("TARGET_WALLET")

# Pagination — small for the example so the loop is exercised on
# realistic portfolios. Production callers can use the API's defaults.
PAGE_LIMIT = 5
MAX_PAGES = 1_000

# ── Setup ──

account = Account.from_key(PRIVATE_KEY)
target_wallet = TARGET_WALLET or account.address


# ── Auth helper (mirrors examples/borrow.py) ──

def sign_request():
    """
    Build EIP-191 auth headers for the credit API. Mirrors the helper in
    examples/borrow.py — keep them in lockstep when the message format
    changes.
    """
    timestamp = str(int(time.time()))
    message = f"Floe Credit API\nTimestamp: {timestamp}"
    signed = account.sign_message(encode_defunct(text=message))
    return {
        "X-Wallet-Address": account.address,
        "X-Signature": "0x" + signed.signature.hex(),
        "X-Timestamp": timestamp,
    }


def get_positions(wallet, *, live=False, include_pending=False, limit=None, skip=None, pending_limit=None):
    """GET /v1/positions/:wallet with optional flags."""
    if not ADDRESS_RE.match(wallet):
        raise ValueError(
            f"Invalid wallet address: {wallet!r}. "
            "Expected a 0x-prefixed 20-byte address."
        )

    params = {}
    if live:
        params["live"] = "true"
    if include_pending:
        params["includePending"] = "true"
    if limit is not None:
        params["limit"] = str(limit)
    if skip is not None:
        params["skip"] = str(skip)
    if pending_limit is not None:
        params["pendingLimit"] = str(pending_limit)

    resp = requests.get(
        f"{API_BASE}/v1/positions/{quote(wallet, safe='')}",
        headers=sign_request(),
        params=params,
        timeout=15,
    )

    if resp.status_code == 503:
        # The API instance was started without an Envio indexer endpoint.
        # getPositions cannot run without it (chain-only fallback would
        # be too slow). Retry against a different instance or surface to
        # the operations team.
        raise RuntimeError(
            f"503 — indexer not configured on {API_BASE}. "
            "Try a different API instance."
        )
    resp.raise_for_status()
    return resp.json()


def fmt_rate(bps_str_or_none):
    if bps_str_or_none is None:
        return "—"
    return f"{int(bps_str_or_none) / 100:.2f}%"


def fmt_health(hf):
    if hf is None:
        return "n/a (indexer mode)"
    if hf >= 2.0:
        tag = "safe"
    elif hf >= 1.2:
        tag = "ok"
    elif hf > 1.0:
        tag = "tight"
    else:
        tag = "LIQUIDATABLE"
    return f"{hf:.3f} ({tag})"


def print_summary(label, body):
    summary = body["summary"]
    print(f"--- {label} ---")
    print(f"  source              : {body['source']}")
    print(f"  ltvStale            : {body['ltvStale']}")
    print(f"  indexerBlockNumber  : {body['indexerBlockNumber']}")
    print(f"  active loans        : {summary['activeLoanCount']}")
    print(f"  total debt          : {int(summary['totalDebt']) / 1e6:,.2f} USDC")
    print(f"  weighted-avg rate   : {fmt_rate(summary['weightedAvgInterestRateBps'])}")
    print(f"  worst health factor : {fmt_health(summary['worstHealthFactor'])}")
    if summary.get("nextMaturityAt"):
        print(f"  next maturity       : {summary['nextMaturityAt']} (unix)")
    print()


def main():
    print(f"Signer:  {account.address}")
    print(f"Target:  {target_wallet}")
    print(f"API:     {API_BASE}")
    print()

    # ── Step 1: indexer-mode read ────────────────────────────────────
    body = get_positions(target_wallet)
    print_summary("Step 1: indexer mode (default)", body)

    # The summary's worstHealthFactor is intentionally null in indexer
    # mode. Calling code that drives liquidation alarms should always
    # use live mode (Step 2).

    # ── Step 2: live-mode read ───────────────────────────────────────
    body = get_positions(target_wallet, live=True)
    print_summary("Step 2: live mode (?live=true)", body)

    # ── Step 3: include pending borrow intents ───────────────────────
    body = get_positions(target_wallet, include_pending=True)
    pending = body.get("pendingBorrowIntents") or []
    print(f"--- Step 3: pending borrow intents ({len(pending)}) ---")
    for intent in pending[:5]:
        print(
            f"  • {intent['offerHash']}  "
            f"borrow={int(intent['borrowAmount']) / 1e6:,.2f} USDC  "
            f"max={fmt_rate(intent['maxInterestRateBps'])}  "
            f"expires={intent['expiry']}"
        )
    if len(pending) > 5:
        print(f"  … and {len(pending) - 5} more")
    print()

    # ── Step 4: paginate ────────────────────────────────────────────
    # Per the API contract, paginate as long as `len(loans) == limit`
    # (the page is "full"). Don't gate on summary.activeLoanCount —
    # that's the wallet-wide total and stays constant across pages.
    #
    # Hard cap on iterations as a safety net: a buggy backend that always
    # returns a full page would otherwise spin forever.
    print(f"--- Step 4: paginate active loans (page size {PAGE_LIMIT}) ---")
    seen = 0
    skip = 0
    pages = 0
    while True:
        if pages >= MAX_PAGES:
            raise RuntimeError(
                f"Pagination exceeded {MAX_PAGES} pages; aborting to avoid an infinite loop."
            )
        page = get_positions(target_wallet, limit=PAGE_LIMIT, skip=skip)
        loans = page["positions"]
        for loan in loans:
            seen += 1
            print(
                f"  loan #{loan['loanId']}: "
                f"debt={int(loan['totalDebt']) / 1e6:,.2f} USDC, "
                f"buffer={fmt_rate(loan['bufferBps'])}"
            )
        pages += 1
        if len(loans) < PAGE_LIMIT:
            break
        skip += PAGE_LIMIT
    print(f"  total surfaced via pagination: {seen}")


if __name__ == "__main__":
    main()
